src/main/kotlin/cacophony/regex/RegexParser.kt

```diff
@@ -0,0 +1,148 @@
+package cacophony.regex
+
+import cacophony.utils.AlgebraicRegex
+import java.util.EmptyStackException
+import java.util.Stack
+
+fun parseRegex(str: String): AlgebraicRegex {
+    val parser = RegexParser()
+    val regex = "($str)"
+    var specialCharacter = false
+    var lastEscaped = -1
+    regex.indices.forEach {
+        val c = regex[it]
+        if (specialCharacter) {
+            if (c !in SPECIAL_CHARACTER_MAP.keys) throw RegexSyntaxErrorException()
+            if (it > 1 && (regex[it - 2] !in "(|" || lastEscaped == it - 2)) parser.pushOperation(ConcatOperator)
+            parser.pushRegex(SPECIAL_CHARACTER_MAP[c]!!) // safe because of previous check
+            specialCharacter = false
+            lastEscaped = it
+        } else {
+            when (c) {
+                '(' -> {
+                    if (it != 0 && (regex[it - 1] !in "(|" || lastEscaped == it - 1)) parser.pushOperation(ConcatOperator)
+                    parser.pushOperation(LeftParenthesis)
+                }
+                ')' -> parser.pushOperation(RightParenthesis)
+                '*' -> parser.pushOperation(StarOperator)
+                '|' -> parser.pushOperation(UnionOperator)
+                '\\' -> specialCharacter = true
+                else -> {
+                    if (regex[it - 1] !in "(|") parser.pushOperation(ConcatOperator)
+                    parser.pushSymbol(c)
+                }
+            }
+        }
+    }
+    return parser.finalize().toAlgebraicRegex()
+}
+
+private class RegexParser {
+    val result = Stack<RegexType>()
```

Comment:
Stack is outdated, use ArrayDeque instead.

Reply:
Done